cmd/end2endtest/helpers.go

```diff
@@ -183,123 +181,65 @@ func (t *e2eElection) waitUntilElectionStarts(electionID types.HexBytes) (*vapi.
+				if csp != nil {
+					return cspGenProof(t.election.ElectionID, voterKey, csp)
+				}
+				return t.api.CensusGenProof(t.election.Census.CensusRoot, voterKey)
+			}()
 			if err != nil {
-				log.Warn(err)
-			}
-			proofCh <- voterProof
-		}
-	}
-
-	pcount := t.config.nvotes / t.config.parallelCount
-	var wg sync.WaitGroup
-	for i := 0; i < len(t.voterAccounts); i += pcount {
-		end := i + pcount
-		if end > len(t.voterAccounts) {
-			end = len(t.voterAccounts)
-		}
-		wg.Add(1)
-		go addNaccounts(t.voterAccounts[i:end], &wg)
-	}
-
-	wg.Wait()
-	log.Debugf("%d/%d sik proofs generated successfully", len(sikProofs), len(t.voterAccounts))
-	stopProofs <- true
-
-	return sikProofs
-}
-
-func (t *e2eElection) generateProofs(root types.HexBytes, isAnonymousVoting bool, csp *ethereum.SignKeys) map[string]*apiclient.CensusProof {
-	type voterProof struct {
-		proof   *apiclient.CensusProof
-		address string
-	}
-	proofs := make(map[string]*apiclient.CensusProof, len(t.voterAccounts))
-	proofCh := make(chan *voterProof)
-	stopProofs := make(chan bool)
-	go func() {
-		for {
-			select {
-			case p := <-proofCh:
-				proofs[p.address] = p.proof
-			case <-stopProofs:
-				return
+				errorChan <- err
 			}
-		}
-	}()
-
-	addNaccounts := func(accounts []*ethereum.SignKeys, wg *sync.WaitGroup) {
-		defer wg.Done()
-		log.Infof("generating %d census proofs", len(accounts))
-		for _, acc := range accounts {
-			voterProof := &voterProof{address: acc.Address().Hex()}
+			accP := acctProof{account: acc, proof: proof}
 
-			var err error
-			if csp != nil {
-				voterProof.proof, err = cspGenProof(t.election.ElectionID, acc.Address().Bytes(), csp)
-			} else {
+			if t.election.VoteMode.Anonymous {
 				voterPrivKey := acc.PrivateKey()
 				voterApi := t.api.Clone(voterPrivKey.String())
-				voterProof.proof, err = voterApi.CensusGenProof(root, acc.Address().Bytes())
+				accP.proofSIK, err = voterApi.GenSIKProof()
 				if err != nil {
-					log.Warn(err)
+					errorChan <- err
 				}
 			}
-			if err != nil {
-				log.Warn(err)
-			}
-
-			if !isAnonymousVoting {
-				voterProof.proof.KeyType = models.ProofArbo_ADDRESS
-			}
-			proofCh <- voterProof
+			t.voters.Store(acc.Public, accP)
+			vcount += 1
 		}
 	}
 
 	pcount := t.config.nvotes / t.config.parallelCount
-	var wg sync.WaitGroup
-	for i := 0; i < len(t.voterAccounts); i += pcount {
+	for i := 0; i < len(voterAccts); i += pcount {
 		end := i + pcount
-		if end > len(t.voterAccounts) {
-			end = len(t.voterAccounts)
+		if end > len(voterAccts) {
+			end = len(voterAccts)
 		}
 		wg.Add(1)
-		go addNaccounts(t.voterAccounts[i:end], &wg)
+		go addNaccounts(voterAccts[i:end], &wg)
 	}
 
-	wg.Wait()
-	log.Debugf("%d/%d census proofs generated successfully", len(proofs), len(t.voterAccounts))
-	stopProofs <- true
+	go func() { // avoid blocking the main goroutine
+		wg.Wait()
+		close(errorChan) // close the error channel after all goroutines have finished
+	}()
```

Comment:
wait... what?

oh that's an interesting idea. the `for err := range errorChan` will stay blocked until the close(errorChan). but then you can ignore my previous comment about the buffered chan (you don't even need the `1`, it can be simply unbuffered)

i like it. but i would describe the mechanism in a comment (and make the errorChan unbuffered, otherwise it looks misleading)

Reply:
like you did in https://github.com/vocdoni/vocdoni-node/pull/1103/files#diff-e41fced2d6ee22c899abb1543729ef493417c53d17c00c7a644be97e6095b3faR355 , that simple comment is enough